Sources/PackageCollectionGenerator/PackageMetadataProviders/GitLabPackageManagerMetadataProvider.swift

```diff
@@ -0,0 +1,170 @@
+                } else if let licenseURL = metadata.license_url,
+                    let licenseURL = URL(string: licenseURL) {
+                    packageLicense = .init(name: nil, url: licenseURL)
+                } else {
+                    packageLicense = nil
+                }
+
+                let model = PackageBasicMetadata(
+                    summary: metadata.description,
+                    keywords: metadata.topics,
+                    readmeURL: metadata.readme_url.flatMap { URL(string: $0) },
+                    license: packageLicense
+                )
+
+                callback(.success(model))
+            default:
+                callback(.failure(Errors.invalidResponse(metadataURL, "Invalid status code: \(response.statusCode)")))
+            }
+        }
+    }
+
+    internal func apiURL(_ url: String) -> Foundation.URL? {
+        return URL(string: url.spm_dropGitSuffix())
+    }
+
+    private func makeRequestOptions(validResponseCodes: [Int]) -> HTTPClientRequest.Options {
+        var options = HTTPClientRequest.Options()
+        options.addUserAgent = true
+        options.validResponseCodes = validResponseCodes
+        options.authorizationProvider = { url in
+            url.host.flatMap { host in
+                return self.authTokens[.gitlab(host)].flatMap { token in "Bearer \(token)" }
+            }
+        }
+        return options
+    }
+
+    private static func makeDefaultHTTPClient() -> HTTPClient {
+        var client = HTTPClient()
+        client.configuration.requestTimeout = .seconds(2)
+        client.configuration.retryStrategy = .exponentialBackoff(maxAttempts: 3, baseDelay: .milliseconds(50))
+        client.configuration.circuitBreakerStrategy = .hostErrors(maxErrors: 50, age: .seconds(30))
+        return client
+    }
+
+    enum Errors: Error, Equatable {
+        case invalidGitURL(URL)
+        case invalidResponse(URL, String)
+        case permissionDenied(URL)
+        case invalidAuthToken(URL)
+        case apiLimitsExceeded(URL, Int, Int)
+        case notFound(URL)
+    }
+}
+
+extension GitLabPackageMetadataProvider {
+    fileprivate struct GetRepositoryResponse: Codable {
+        let name: String
+        let fullName: String
+        let description: String?
+        let topics: [String]?
+        let license_url: String?
+        let readme_url: String?
+        let license: License?
+
+        private enum CodingKeys: String, CodingKey {
+            case name
+            case fullName = "name_with_namespace"
+            case description
+            case topics
+            case license_url
+            case readme_url
+            case license
+        }
+    }
+}
+
+extension GitLabPackageMetadataProvider {
+    fileprivate struct License: Codable {
+        let htmlURL: Foundation.URL
+        let sourceURL: Foundation.URL
+        let nickname: String?
+        let name: String?
```

Comment:
We need to add `key` here since that's what we would use to populate the package collection.